src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -547,7 +577,62 @@ def _initialize(self) -> None:
         self._engine = create_engine(
             url=url, connect_args=connect_args, **engine_args
         )
-        SQLModel.metadata.create_all(self._engine)
+        self._alembic = Alembic(self.engine)
+        if (
+            not self.skip_migrations
+            and ENV_ZENML_DISABLE_DATABASE_MIGRATION not in os.environ
+        ):
+            self.migrate_database()
+
+    def migrate_database(self) -> None:
+        """Migrate the database to the head as defined by the current python package."""
+        alembic_logger = logging.getLogger("alembic")
+
+        # remove all existing handlers
+        while len(alembic_logger.handlers):
+            alembic_logger.removeHandler(alembic_logger.handlers[0])
+
+        logging_level = get_logging_level()
+
+        # suppress alembic info logging if the zenml logging level is not debug
+        if logging_level == LoggingLevels.DEBUG:
+            alembic_logger.setLevel(logging.DEBUG)
+        else:
+            alembic_logger.setLevel(logging.WARNING)
+
+        alembic_logger.addHandler(get_console_handler())
+
+        # We need to account for 3 distinct cases here:
+        # 1. the database is completely empty (not initialized)
+        # 2. the database is not empty, but has never been migrated with alembic
+        #   before (i.e. was created with SQLModel back when alembic wasn't
+        #   used)
+        # 3. the database is not empty and has been migrated with alembic before
+
+        revisions = self.alembic.current_revisions()
+        if len(revisions) >= 1:
+            if len(revisions) > 1:
+                logger.warning(
+                    "The ZenML database has more than one migration head "
+                    "revision. This is not expected and might indicate a "
+                    "database migration problem. Please raise an issue on "
+                    "GitHub if you encounter this."
+                )
+            # Case 3: the database has been migrated with alembic before. Just
+            # upgrade to the latest revision.
+            self.alembic.upgrade()
+        else:
+            if self.alembic.db_is_empty():
```

Comment:
the conditionals here could perhaps be flattened a bit via `elif` and/or early returns